src/wallet/rpcdump.cpp

```diff
@@ -1149,3 +1149,51 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
 
     return response;
 }
+
+UniValue rescanblockchain(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp))
+        return NullUniValue;
+
+    if (request.fHelp || request.params.size() > 2)
+        throw std::runtime_error(
+            "rescanblockchain (\"start-height\") (\"stop-height\")\n"
+            "\nRescan the local blockchain for wallet related transactions.\n"
+            "\nArguments:\n"
+            "1. \"start-height\"    (number, optional) blockheight where the rescan should start\n"
+            "2. \"stop-height\"     (number, optional) blockheight where the rescan should stop\n"
+            "\nExamples:\n"
+            + HelpExampleCli("rescanblockchain", "\"100000 120000\"")
+            + HelpExampleRpc("rescanblockchain", "\"100000 120000\"")
+            );
+
+    LOCK2(cs_main, pwallet->cs_wallet);
+
+    CBlockIndex *pIndexStart = NULL;
+    CBlockIndex *pIndexStop = NULL;
+    if (request.params.size() > 0 && request.params[0].isNum())
+        pIndexStart = chainActive[request.params[0].get_int()];
+
+    if (request.params.size() > 1 && request.params[1].isNum())
+        pIndexStop = chainActive[request.params[1].get_int()];
+
+    if (!pIndexStart)
+         pIndexStart = chainActive.Genesis();
+
+    //We can't rescan beyond non-pruned blocks, stop and throw an error
+    if (fPruneMode)
+    {
+        CBlockIndex *block = chainActive.Tip();
+        while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA))
+            block = block->pprev;
+
+        if (pIndexStart->nHeight < block->nHeight)
+            throw JSONRPCError(RPC_WALLET_ERROR, "Can't rescan beyond pruned data.");
+    }
+
+    if (pwallet)
+        pwallet->ScanForWalletTransactions(pIndexStart, pIndexStop, true);
+
+    return NullUniValue;
```

Comment:
Should at least throw an error this isn't successful (the ScanForWalletTransactions call will return null on success, otherwise a pointer to the last failing block, so this should be pretty easy).